prov/lnx/include/lnx.h

```diff
@@ -0,0 +1,585 @@
+
+int lnx_fabric_close(struct fid *fid);
+
+int lnx_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		    struct fid_domain **dom, void *context);
+
+int lnx_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
+		struct fid_av **av, void *context);
+
+int lnx_cq_open(struct fid_domain *domain, struct fi_cq_attr *attr,
+		struct fid_cq **cq, void *context);
+
+int lnx_endpoint(struct fid_domain *domain, struct fi_info *info,
+		 struct fid_ep **ep, void *context);
+
+int lnx_scalable_ep(struct fid_domain *domain, struct fi_info *info,
+		    struct fid_ep **ep, void *context);
+
+int lnx_cq2ep_bind(struct fid *fid, struct fid *bfid, uint64_t flags);
+
+int lnx_get_msg(struct fid_peer_srx *srx, struct fi_peer_match_attr *match,
+		struct fi_peer_rx_entry **entry);
+int lnx_get_tag(struct fid_peer_srx *srx, struct fi_peer_match_attr *match,
+		struct fi_peer_rx_entry **entry);
+int lnx_queue_msg(struct fi_peer_rx_entry *entry);
+int lnx_queue_tag(struct fi_peer_rx_entry *entry);
+void lnx_free_entry(struct fi_peer_rx_entry *entry);
+void lnx_foreach_unspec_addr(struct fid_peer_srx *srx,
+	fi_addr_t (*get_addr)(struct fi_peer_rx_entry *));
+
+static inline struct lnx_peer *
+lnx_get_peer(struct lnx_peer **peers, fi_addr_t addr)
+{
+	if (!peers || addr == FI_ADDR_UNSPEC)
+		return NULL;
+
+	return peers[addr];
+}
+
+static inline
+void lnx_get_core_desc(struct lnx_mem_desc *desc, void **mem_desc)
+{
+	if (desc && desc->desc[0].core_mr) {
+		if (mem_desc)
+			*mem_desc = desc->desc[0].core_mr->mem_desc;
+		return;
+	}
+
+	*mem_desc = NULL;
+}
+
+static inline
+int lnx_create_mr(const struct iovec *iov, fi_addr_t addr,
+		  struct lnx_domain *lnx_dom, struct fid_mr **fid_mr)
+{
+	struct ofi_mr *mr;
+	struct ofi_mr_entry *mre;
+	struct fi_mr_attr attr = {};
+	struct fi_mr_attr cur_abi_attr;
+	struct ofi_mr_info info = {};
+	uint64_t flags;
+	int rc;
+
+	attr.iov_count = 1;
+	attr.mr_iov = iov;
+	mre = ofi_mr_cache_find(&lnx_dom->ld_mr_cache, &attr, 0);
+	if (mre) {
+		mr = (struct ofi_mr *)mre->data;
+		goto out;
+	}
+
+	attr.iface = ofi_get_hmem_iface(iov->iov_base,
+			&attr.device.reserved, &flags);
+	info.iov = *iov;
+	info.iface = attr.iface;
+	rc = ofi_hmem_dev_register(attr.iface, iov->iov_base, iov->iov_len,
+				   (uint64_t *) &attr.hmem_data);
+	if (rc)
+		return rc;
+
+	rc = ofi_mr_cache_search(&lnx_dom->ld_mr_cache, &info, &mre);
+	if (rc)
+		return rc;
+
+	mr = (struct ofi_mr *)mre->data;
+	ofi_mr_update_attr(lnx_dom->ld_domain.fabric->fabric_fid.api_version,
+			   lnx_dom->ld_domain.info_domain_caps, &attr, &cur_abi_attr, 0);
+
+	mr->mr_fid.fid.fclass = FI_CLASS_MR;
+	mr->mr_fid.fid.context = attr.context;
+	mr->domain = &lnx_dom->ld_domain;
+	mr->flags = flags;
+	mr->iface = cur_abi_attr.iface;
+	mr->device = cur_abi_attr.device.reserved;
+	mr->hmem_data = cur_abi_attr.hmem_data;
+	mr->mr_fid.mem_desc = (void*) mr;
+
+out:
+	*fid_mr = &mr->mr_fid;
+	ofi_mr_cache_delete(&lnx_dom->ld_mr_cache, mre);
```

Comment:
As far as I can understand the caching code, that doesn't actually remove it from the cache. it adds it to the lru list. Only if the cache is full and you're trying to add more entries (using the search API) then the lru_list is flushed. Otherwise it remains there and is found in the next search operation.

Reply:
It also returns a reference for the MR cached entry making it eligible to be freed. My concern here is that you added a MR reference via ofi_mr_cache_search/find(), gave the handle to the caller of lnx_create_mr(), and then returned the reference while the caller is still using the MR. ofi_mr_cache_delete() should only be called when the MR is no longer needed.

Reply:
good point. fixed. Note this area will be visited in a follow up PR to generalize the mr caching so the structure of the code will change.